### ISSUE TITLE
tf.feature_column.input_layer do not well support dtype=float64, therefore the highlevel API

### ISSUE THREAD

**Issue by bewantbe** [tensorflow/tensorflow#20358]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No.

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Debian 9.4.

- **TensorFlow installed from (source or binary)**:
source.

- **TensorFlow version (use command below)**:
1.9.0-rc1

- **Python version**: 
3.5.3

- **Bazel version (if compiling from source)**:
0.11.1

- **GCC/Compiler version (if compiling from source)**:
6.3.0

- **CUDA/cuDNN version**:
NA (cpu only build)

- **GPU model and memory**:
NA (cpu only build)

- **Exact command to reproduce**:
See attached python script. Just run it like:
`python3 test_tf_float64_min.py`

### The problem
The function `tf.feature_column.input_layer` return tensor in dtype=float32 for input dtype=float64.

I know this is a [documented](https://github.com/tensorflow/tensorflow/blob/v1.9.0-rc1/tensorflow/python/feature_column/feature_column.py#L270) behavior, but this breaks things:

1. When the training data is in float64 precision, the user is (I am) expecting a float64 output, especially when the DNN written in high level API is used as a regressor. However, it returns float32, see the script in the end.

2. When building a custom estimator using high level API, seems that the input layer has to be created by `tf.feature_column.input_layer`. See the guide [here](https://www.tensorflow.org/get_started/custom_estimators#define_the_input_layer). This forbid the user from creating a network with dtype=float64 precision.

One may argue that dtype=float32 is already more than enough for DNN. But e.g. when doing theoretical work, the extra precision is desired.

A possible fix is:

Change line [2347](https://github.com/tensorflow/tensorflow/blob/v1.9.0-rc1/tensorflow/python/feature_column/feature_column.py#L2347) of function `_transform_feature` in file `tensorflow/python/feature_column/feature_column.py`, from

```python3
 return math_ops.to_float(input_tensor)
```

to

```python3
    if not input_tensor.dtype.is_floating:
      return math_ops.to_float(input_tensor)
    return input_tensor
```

Note: this fix could breaks old code that has float32 assumption...

If this is considered as non-bug, then consider this as a feature request that make it possible to pass dtype=float64 through.

### Source code / logs
```python3
# Test tensorflow high level API.
# dtype=float64 will be silently transformed to float32.

import sys
import numpy as np
import tensorflow as tf

def load_data(dtype):
    l = 7
    x1 = np.linspace(-1.0, 1.0, l, dtype=dtype)
    z1 = np.sin(x1)
    return ({'x':x1}, {'z':z1})

def train_input_fn(features, labels, batch_size):
    dataset = tf.data.Dataset.from_tensor_slices((features, labels['z']))
    dataset = dataset.repeat().batch(batch_size)
    return dataset

def eval_input_fn(features):
    dataset = tf.data.Dataset.from_tensor_slices(features)
    dataset = dataset.batch(1)
    return dataset

def main(argv):
    dtype = np.float64   # specifies the data type

    batch_size = 7
    n_steps = 1

    (train_x, train_y) = load_data(dtype)
    my_feature_columns = [
        tf.feature_column.numeric_column(
            key   = 'x',
            shape = (1,),
            dtype = tf.as_dtype(train_x['x'].dtype))
    ]

    regressor = tf.estimator.DNNRegressor(   # or any other regressors
        hidden_units    = [3],
        feature_columns = my_feature_columns)

    regressor.train(
        input_fn = lambda:train_input_fn(train_x, train_y, batch_size),
        steps    = n_steps)

    predict_x = {'x': np.array([0.1, 0.2])}

    predictions = regressor.predict(
        input_fn=lambda:eval_input_fn(predict_x))

    print([i for i in predictions]) # outputs are float32 instead of float64

if __name__ == '__main__':
    tf.app.run(main)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 32 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 48 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 63 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 78 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 93 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 108 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 123 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by pdxjohnny** [comment]
After the switch from 1.X regressor to 2.X regressor we're seeing this as well. We're also getting terrible accuracy now and not sure if its related.

**Comment by sushreebarsa** [comment]
@bewantbe It looks like you are using an older Version of Tensorflow (1.x). Many bugs have been fixed in the latest version. Can you please execute your code using Latest Version (2.6.0) and let us know if the issue still persists? Please refer to the [link](https://www.tensorflow.org/api_docs/python/tf/compat/v1/feature_column/input_layer) and let us know if it helps ? Thanks!
